### PR TITLE
Integrating PPS full sim in Run3 (only)

### PR DESCRIPTION
**Pull request by mundim**
#### PR description:

Another step towards the ntegration of  PPS full simulation  into cmssw (still only for Run3).

This PR propagates the changes made by @civanch in the central simulation. Now, PPS simulation is triggered by the flag LHCTransport (False by default), which needs to be set to TRUE.  Since the PPS geometry is not yet loaded into the DB, the digi was removed from the chain in order to avoid failure in the standard tests. In summary, this PR changes SimPPS/PPSSimTrackProducer and PPSProtonTransport to save inthe LHCTransport data container only the propagate protons (only secondary vertices in from of PPS region). 

Another PR will be issued when PPS geometry gets loaded into the DB, pluging in the digi part.

#### PR validation:

scram b runtests ran with no issue
scram b code-checks  and scram b code-formats also ran ok 

runTheMatrix.py -l limited -i all --ibeos was ran again with the results:
35 33 27 20 15 4 1 1 1 tests passed, 0 1 5 0 0 0 0 0 0 failed

All the 6 wf failed due to timeout. Running them again gave:
6 6 6 5 2 tests passed, 0 0 0 0 0 failed

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

Before submitting your pull requests, make sure you followed this checklist:
- verify that the PR is really intended for the chosen branch
- verify that changes follow [CMS Naming, Coding, And Style Rules](http://cms-sw.github.io/cms_coding_rules.html)
- verify that the PR passes the basic test procedure suggested in the [CMSSW PR instructions](https://cms-sw.github.io/PRWorkflow.html)

### CODE REVIEW

File: SimG4Core/Application/python/g4SimHits_cfi.py [hunk 1]
```diff
@@ -575,6 +575,23 @@
 run2_common.toModify( g4SimHits.HFShowerLibrary, FileName = 'SimG4CMS/Calo/data/HFShowerLibrary_npmt_noatt_eta4_16en_v4.root' )
 run2_common.toModify( g4SimHits.HFShower, ProbMax = 0.5)
 
+
+from Configuration.Eras.Modifier_ctpps_2021_cff import ctpps_2021
```
Comment (civanch) [inline SimG4Core/Application/python/g4SimHits_cfi.py:579]: @mundim , if the modifier is configured in this way, then central detector simulation is disabled and only PPS simulation is enabled. In that case, this modifier should be applied only for specific ctpps_2021 simulation. I assume it is the case and in the normal WF used in tests this should not be used.

**Comment by civanch** [comment]
@mundim , rebase is needed.

File: SimG4Core/Application/python/g4SimHits_cfi.py (hunk 1 above)
Comment (mundim) [reply SimG4Core/Application/python/g4SimHits_cfi.py:579]: Hi @civanch (sorry for the late reply, my laptop had to be taken to the maintenance, I'm using a tablet).

I'm not sure I understood the problem. I just picked up from the comments in the PR 29916, as suggested by @makortel for Run2  (https://github.com/cms-sw/cmssw/pull/29916#discussion_r428796936) and replaced ctpps_2016 by ctpps_2021.

Do I have to make others changes or the rebase as your next message is enough?

Comment (makortel) [reply SimG4Core/Application/python/g4SimHits_cfi.py:579]: @civanch Could you clarify in which way the customization below disables central detector simulation? Is it perhaps the `g4simhits.Generator.HepMCProductLabel = 'LHCTransport`?

Comment (mundim) [reply SimG4Core/Application/python/g4SimHits_cfi.py:579]: I don' t believe so. We have done private pps simulation with this label and the central CMS simulation was not disabled.

**Comment by civanch** [comment]
please test

**Comment by civanch** [comment]
@mundim , I have run this PR on top of the recent IB and can confirm that in this form full simulation is not OK. 3 times more CPU is required for MinBias events. I will continue investigation but the main source of the problem come from extended eta cut for all primary particles. I think, we need re-discuss how PPS simulation is enabled. The main approach should be not change parameters used for central detector but only add PPS particles which should cost negligible addition.

**Comment by civanch** [comment]
@mundim , trying to understand what is going on I found out several problematic printouts and make #30652. Inside this PR I hope that the check of eta for beam particles now disabled, so no need to extend eta cut is required. A second moment: time cut. As I said before, it may be removed transparently if PPS detectors will be inside PPS G4Region. This G4Region may defined in the xml geometry together with cuts. If the region is defined, then in g4SimHits we may define time cut specifically to PPS. Finally we may enable LHC transport by default, so no modifier will be required for g4SimHits_cfi.py

**Comment by mundim** [comment]
> @mundim , I have run this PR on top of the recent IB and can confirm that in this form full simulation is not OK. 3 times more CPU is required for MinBias events. I will continue investigation but the main source of the problem come from extended eta cut for all primary particles. I think, we need re-discuss how PPS simulation is enabled. The main approach should be not change parameters used for central detector but only add PPS particles which should cost negligible addition.

Hi @civanch , I' m afraid the same problem we had with CASTOR is happening, that required the developmento of a shower library and GFlash. Ianna has mentioned in the beginning of this integration long about the need of a dedicated PPS region, but it wa posponed to after the integration, which never occured so far.

The only particle we need for the extended eta range is the scattered proton. The other particles  are just ignored. What about if we kill and remove them from the stack in such a way

**Comment by mundim** [comment]
> @mundim , trying to understand what is going on I found out several problematic printouts and make #30652. Inside this PR I hope that the check of eta for beam particles now disabled, so no need to extend eta cut is required. A second moment: time cut. As I said before, it may be removed transparently if PPS detectors will be inside PPS G4Region. This G4Region may defined in the xml geometry together with cuts. If the region is defined, then in g4SimHits we may define time cut specifically to PPS. Finally we may enable LHC transport by default, so no modifier will be required for g4SimHits_cfi.py

I' m not really expert on this g4SimHits stuff, it is likely you have a muc better solution, but without the eta extension, the simhit of pps were not produced. You need anything else I have to implement and test, could you please send me the intructions?

**Comment by civanch** [comment]
@mundim , I hope, that with #30652 you do not need to change eta cut anymore. Let us try this out.

**Comment by civanch** [comment]
@mundim , I would propose exclude g4Sim_Hits modifications from this PR and resolve the issue separately.

**Comment by mundim** [comment]
> @mundim , I would propose exclude g4Sim_Hits modifications from this PR and resolve the issue separately.

Hi @civanch ,  I guess I need to wait for PR #30652 to be merged to process, right?

**Comment by mundim** [comment]
> @mundim , I would propose exclude g4Sim_Hits modifications from this PR and resolve the issue separately.

Do you mean only the eta range extension or also the HepMCProductLabel?

**Comment by silviodonato** [comment]
> > @mundim , I would propose exclude g4Sim_Hits modifications from this PR and resolve the issue separately.
> 
> Hi @civanch , I guess I need to wait for PR #30652 to be merged to process, right?

#30652 moved to https://github.com/cms-sw/cmssw/pull/30887

**Comment by silviodonato** [comment]
@mundim  #30887 ( it was #30652 ) has been merged. Tomorrow we will cut `CMSSW_11_2_0_pre3`

**Comment by civanch** [comment]
@mundim , you may include in this PR the file SimG4Core/Application/python/g4SimHits_cfi.py from the recent IB and let us try  out how this will work.

**Comment by mundim** [comment]
Hi @civanch , I; ll try it right now. thanks

**Comment by mundim** [comment]
> @mundim , you may include in this PR the file SimG4Core/Application/python/g4SimHits_cfi.py from the recent IB and let us try out how this will work.

Hello @civanch,  I tried removing the extension of the eta range but it didn't work. There is no hit produced in PPS. You mention about PPS region. Long ago Ianna has said it would be advisable to create such a region but it was postponed, so I guess we need to do it now. I have no idea about how to do it though.

Another temporary solution could be to remove all particles in the extended eta range excepting the two scattered protons. Do you think this could work?

**Comment by mundim** [comment]
I know ZDC has a dedicated region, maybe I can have a look on it and do it similarly. Do you have any advice?

**Comment by civanch** [comment]
@mundim , ZDC is a simple example and PPS should use something similar defining G4Region.

Concerning failing of the test: you mean tests inside SimPPS/Configuration/tests or 11634.0?

**Comment by mundim** [comment]
> @mundim , ZDC is a simple example and PPS should use something similar defining G4Region.
> 
> Concerning failing of the test: you mean tests inside SimPPS/Configuration/tests or 11634.0?
I mean SimPPS/Configuration/tests; to have the PPS hits produced the eta range extension and the increase of maxtime are needed.

**Comment by mundim** [comment]
I ran again the tests (still with the eta extension) and the error due to the db tag for gains and daqmap disapperared, however the tests is still failing because the input from step1 was produced with release 106X in which the pps geometry was not present. How can I run the tests forcing the step1.root to be proeduced with the current release?

**Comment by perrotta** [comment]
> I ran again the tests (still with the eta extension) and the error due to the db tag for gains and daqmap disapperared, however the tests is still failing because the input from step1 was produced with release 106X in which the pps geometry was not present. How can I run the tests forcing the step1.root to be produced with the current release?

Which is the test that fails? Normally step1 is produced from scratch in the standard relval workflows.
On the other hand, the code should still be able to read past formats: I think that some switch should be allowed in case old gen-sim files are given in input.

**Comment by mundim** [comment]
> > I ran again the tests (still with the eta extension) and the error due to the db tag for gains and daqmap disapperared, however the tests is still failing because the input from step1 was produced with release 106X in which the pps geometry was not present. How can I run the tests forcing the step1.root to be produced with the current release?
> 
> Which is the test that fails? Normally step1 is produced from scratch in the standard relval workflows.
> On the other hand, the code should still be able to read past formats: I think that some switch should be allowed in case old gen-sim files are given in input.

The  11634.0 and 12434.0, as reported at the beginning... these tests fetch the step1.root from DAS with dataset=/RelValTTbar_14TeV/CMSSW_10_6_1-106X_mcRun3_2021_realistic_v1_rsb-v1/GEN-SIM

It reads the file, but PPS Pixel hits collection (PSimHits_g4SimHits_CTPPSPixelHits_SIM) doesn't exist in this file. The only solution I see to allow the process of previously created step1.root is to protect the iEvent.getByToken call.

**Comment by mundim** [comment]
> @mundim , ZDC is a simple example and PPS should use something similar defining G4Region.
> 
> Concerning failing of the test: you mean tests inside SimPPS/Configuration/tests or 11634.0?

@civanch, you mention in the simulation meeting I could use the same G4Region already defined in the PPS XML file at Geometry/VeryForwardData/data/CTPPS_Cuts_Per_Region.xml. There are 4 regions defined: 
RPSiliconDetectorsCuts
RPRomanPotParts
RPOtherDetPartsCut
RP_NonDenseMaterialsCuts

To begin with, I tried to change the MaxTrackTime as defined in g4SimHits_cfi.py at common_maximum_time, but it worked only partially, apparently these regions doesn't cover all the PPS region.

**Comment by mundim** [comment]
In time... I found another region like object, defined at Geometry/VeryForwardData/data/CTPPS_Pixel_Sens.xml, called CTPPSPixelRegion. I changed g4SimHits_cfi from:
```
common_maximum_time = cms.PSet(
    MaxTrackTime  = cms.double(500.0),
    MaxTimeNames  = cms.vstring('ZDCRegion')
   MaxTrackTimes = cms.vdouble(2000.0)
   ...
```

to

```
common_maximum_time = cms.PSet(
    MaxTrackTime  = cms.double(500.0),
    MaxTimeNames  = cms.vstring('ZDCRegion','CTPPSPixelRegion'),
    MaxTrackTimes = cms.vdouble(2000.0,2000.0),
    ...
```
But it didn't work. I guess a new region should be created or mofidy one of the above. Any suggestion? I would say to create a new one to avoid clash with the already debugged geometry.

**Comment by civanch** [comment]
@mundim , it is a correct proposal.

**Comment by mundim** [comment]
> @mundim , it is a correct proposal.

Is there any documentation about the G4Region I can have a look? I googled it but found nothing so far. Looking into the zdc one didn't help too much. Thank you.

**Comment by mundim** [comment]
Hi

I played with some examples to write a PPS region but I could not get a clear picture. However, I played with the existing PPS regions and I could get some conclusion. As I said, PPS has some prodcuts regions already defined, but not of them could do the job alone, I had to include (almost) all of them to get the same hits as before. I the study I've done there is still the extension of the eta range, let's discuss it down below. It was clear there is two things we need to do to get the PPS hits: remove the eta cut (so far, extending the Eta cut to ±13) AND increasing the MaxTrackTime in the StackingAction and SteppingAction. As for these last one, I went further in the path as reported above, including the other pps regions:

```
common_maximum_time = cms.PSet(
    MaxTrackTime  = cms.double(500.0),
MaxTrackNames=cms.vstring('ZDCRegion','RPSiliconDetectorsCuts','RPRomanPotParts','RP_NonDenseMaterialsCuts','CTPPSPixelRegion','CTPPSPixelSensitiveRegion'),
    MaxTrackTimes = cms.vdouble(2000.0,2000.0,2000.0,2000.0,2000.0,2000.0),
(...)
```
then I got basically the same hits as the global increase of the tracking time. The question now is: can we include all these regions in the g4SimHits_cfi.py as above? I'm in favor of raving a single region for this purpose. The question is "how?" I'll see the documentation sent by email but if some of you have any advice to speed it up I'll appreciate it.

The second question is how to avoid eta cut for the scattered protons without extending the eta range globally.

**Comment by civanch** [comment]
@mundim , I am sure, that 4 regions for PPS is an overhead. It is enough to have only one and to set set cuts 0.1 mm for all PPS volumes. Because these regions are already defined we can agree to keep as it is. It is possible to have these 4 regions but no need to have different cuts.

**Comment by civanch** [comment]
@mundim , what is essential - all PPS volumes should be inside PPS regions in order to have the same time window.

**Comment by civanch** [comment]
Concerning eta cut: I believe, that now forward protons are not supressed by eta cut.

**Comment by mundim** [comment]
> @mundim , what is essential - all PPS volumes should be inside PPS regions in order to have the same time window.

When I started to develop the simulation geometry, I noticed a tube from ZDC to FP420 in the cms extended geometry. Then I included a "cut" on that tube going from ZDC to (old)HPS, then from HPS to FP420 and a tube to host the PPS volumes, but there were some discussion in the group and this idea was abandoned. My impression is that there is not a single PPS volume (envelop) to hold all the volumes. I'll take a more careful look on the geometry.

I have a question though: these 4 (or 6 regions) have different cuts, from 0.1mm to 10mm. What would be the consequence of having a region with different cuts than the ones already existing if they are inside the newly created one?

**Comment by mundim** [comment]
> Concerning eta cut: I believe, that now forward protons are not supressed by eta cut.

Do you mean the latest IB? Is it already available?

**Comment by civanch** [comment]
@mundim , there is no reason to set 10 mm cut for PPS devices. Such cuts are defined in air or in thick absorbers like concrete. for accurate simulation of a signal in Silicon sensors usual recommendation is 1/3 of sensitive detector thickness. In PPS there is no thick elements, and no CPU problems, so cuts for sensitive elements and for support structure may be identical.

**Comment by mundim** [comment]
> @mundim , there is no reason to set 10 mm cut for PPS devices. Such cuts are defined in air or in thick absorbers like concrete. for accurate simulation of a signal in Silicon sensors usual recommendation is 1/3 of sensitive detector thickness. In PPS there is no thick elements, and no CPU problems, so cuts for sensitive elements and for support structure may be identical.

This is already defined there since long ago (coming from the totem simulation). I would not like to change it now without a careful debuging. I believe that this is in the region between the stations where detailed simulaition is not needed.

**Comment by civanch** [comment]
@mundim , thank you for these comments, now I see an extra problem, which may be addressed differently and I will need to make a new PR which will exclude time cut check in very forward area of CMS. I will try to exclude time cut for ZDC and PPS at all, 

Problem is in the trajectory of scattered protons - they path beam pipe or may pass elements, which are not PPS. Due to this potential situation, such protons may be killed due to time cut in unexpected place.

Concerning your questions: 

1) G4Region is not necessary a continues object - one G4Region may have several "root logical volumes", which are far away from each other.

2) High cuts in air or not important volumes may be left unchanged.

3) PR, in which protons are not checked for eta cuts was merged about 2 weeks ago.

**Comment by mundim** [comment]
>     3. PR, in which protons are not checked for eta cuts was merged about 2 weeks ago.
I checked out the latest IB but without the eta range extension, there is no simvertex at PPS region. As I mentioned there are two requirements: 
1 - the extension of the eta range (or whatever other solution) in order to have the simvertex created at the starting of PPS region
2 - the increase of the MaxTime to have the simhits in PPS.

Point 1 above should have been solved according to what you said, but it didn't.

**Comment by mundim** [comment]
Hello... Just an update. The changes implemented in PR #31157 solved both the problems of eta range and maxtime, but it is still needed to change the label of HepMCProduct to LHCTranporst and this is causing a huge increase in processing time. I'm already talking to @civanch to find a solution.

**Comment by mundim** [comment]
This commit  propagates the changes made by @civanch in the central simulation. Now, PPS simulation is triggered by the flag LHCTransport, which needs to be set to TRUE. It was realized that the failure in wf 11634.0 and 12434.0 was due to the missing geometry in the DB, then the tests were fetching the GEN+SIM file from a simulation done with CMSSW_10_6_1, with missing PPS digis. So, the digi part was removed from the chain until PPS geometry is loaded into the DB, when another PR will be issued with only a couple of lines of code changed.

**Comment by perrotta** [comment]
@cmsbuild please test

**Comment by civanch** [comment]
+1

@mundim , please, update initial description, since PR was proposed for the first time some modifications were added.

**Comment by mundim** [comment]
> +1
> 
> @mundim , please, update initial description, since PR was proposed for the first time some modifications were added.

Tell me if it is ok now. thanks.

**Comment by smorovic** [comment]
+1

**Comment by perrotta** [comment]
+1
- ctppsPixelRawData included in the ctppsRawData Task for Run3
- digi's still left as placeholders, while waiting for an updated PPS  geometry in the DB
- Jenkins tests pass and show no differences in reco outputs

**Comment by perrotta** [comment]
@mundim the static analyzer shows that `double theta` is uselessly defined (L25) and modified (L35-36) in SimTransport/PPSProtonTransport/src/BaseProtonTransport.cc

It could be cleaned either here, or in the next PR, if you have to touch the same file

**Comment by mundim** [comment]
> @mundim the static analyzer shows that `double theta` is uselessly defined (L25) and modified (L35-36) in SimTransport/PPSProtonTransport/src/BaseProtonTransport.cc
> 
> It could be cleaned either here, or in the next PR, if you have to touch the same file

If you don't mind, I prefer to leave it to the next PR (which will be very small, I mean,  very few changes). This one has already a huge history, any change will restart it again. What do you think? And still, it is planned to change the transport again soon, when realistic optics will be available for run3, so we will have at least two opportunities to address this problem.

**Comment by perrotta** [comment]
> > @mundim the static analyzer shows that `double theta` is uselessly defined (L25) and modified (L35-36) in SimTransport/PPSProtonTransport/src/BaseProtonTransport.cc
> > It could be cleaned either here, or in the next PR, if you have to touch the same file
> 
> If you don't mind, I prefer to leave it to the next PR (which will be very small, I mean, very few changes). This one has already a huge history, any change will restart it again. What do you think? And still, it is planned to change the transport again soon, when realistic optics will be available for run3, so we will have at least two opportunities to address this problem.

Sure, I only wanted to point it to you, in case you missed it!

**Comment by mundim** [comment]
> > > @mundim the static analyzer shows that `double theta` is uselessly defined (L25) and modified (L35-36) in SimTransport/PPSProtonTransport/src/BaseProtonTransport.cc
> > > It could be cleaned either here, or in the next PR, if you have to touch the same file
> > 
> > 
> > If you don't mind, I prefer to leave it to the next PR (which will be very small, I mean, very few changes). This one has already a huge history, any change will restart it again. What do you think? And still, it is planned to change the transport again soon, when realistic optics will be available for run3, so we will have at least two opportunities to address this problem.
> 
> Sure, I only wanted to point it to you, in case you missed it!

thanks... I'll do it right away and leave it in git for the proper time to submit.  I'll need to scan all PR's because there are other issues that were left behind for "future PR". As soon as we get this integrated, it will be time to address them all.

**Comment by civanch** [comment]
@makortel , @Dr15Jones , @silviodonato , @qliphy , please check and sign this PR - it stay too long before can be merged, now is in a good shape and include needed modifications for PPS simulation. Following PR will be needed, when GT with full Run-3 geometry will be available.

**Comment by silviodonato** [comment]
+operations

**Comment by silviodonato** [comment]
Kind reminder @cms-sw/core-l2

File: IOMC/RandomEngine/python/IOMC_cff.py [hunk 2]
```diff
@@ -19,7 +19,7 @@
     ),
     LHCTransport = cms.PSet(
         initialSeed = cms.untracked.uint32(87654321),
-        engineName = FastSimEngine
+        engineName = cms.untracked.string('TRandom3')
```
Comment (makortel) [inline IOMC/RandomEngine/python/IOMC_cff.py:22]: Do we really want to use `TRandom3` engine? @civanch

**Comment by makortel** [comment]
+1

Thanks @civanch for the explanation.

**Comment by silviodonato** [comment]
+1

**Pull request merged**